Lib/_pyrepl/unix_console.py

```diff
@@ -417,10 +427,25 @@ def wait(self, timeout: float | None = None) -> bool:
         """
         Wait for events on the console.
         """
-        return (
-            not self.event_queue.empty()
-            or bool(self.pollob.poll(timeout))
-        )
+        if not self.event_queue.empty():
+            return True
+
+        current_thread = threading.current_thread()
+
+        if self._polling_thread is current_thread:
+            # This is a re-entrant call from the same thread
+            # like old repl runtime error
+            raise RuntimeError("can't re-enter readline")
+
+        if not self._poll_lock.acquire(blocking=False):
+            return False
+
+        try:
+            self._polling_thread = current_thread
+            return bool(self.pollob.poll(timeout))
+        finally:
+            self._polling_thread = None
+            self._poll_lock.release()
```

Comment:
```suggestion
        if not self.event_queue.empty():
            return True
        current_thread = threading.current_thread()
        if self._polling_thread is current_thread:
            # Forbid re-entrant calls and use the old REPL error message.
            raise RuntimeError("can't re-enter readline")
        if not self._poll_lock.acquire(blocking=False):
            return False
        try:
            self._polling_thread = current_thread
            return bool(self.pollob.poll(timeout))
        finally:
            self._polling_thread = None
            self._poll_lock.release()
```

We tend to be compact.